In [2]:
import pandas as pd

In [3]:
df = pd.read_csv(r"C:\Users\HP i5\Desktop\SE_Factory\Week 2\PYTHON\WA_Fn-UseC_-Telco-Customer-Churn.csv")

In [4]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

After observing this dataset, we can understand that it contains information about customers using phone and internet services. Each columns shows different details about the customer. The main focus of the dataset seems to be on whether the customer churned (stopped using the service) and what factors influence them to do that. Hence, this dataset helps us find out why customers leave the service and what makes them do that. 

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


This function provides us with information about the dataset:
1. The dataset consists of 7043 rows
2. There are total of 21 columns (Attributes)
3. There are no missing values in the dataset 
4. The dataset contains different data types (object, float64, int64) which reflect the nature of the values in each column.
5. Customer id is the unique identifier of each customer.
The columns with object values contain textual information about the attributes or Yes/No about the availability of the attribute for each customer.

The senior citizen column is a categorical variable (binary values between 1 to 0). 
Tenure as an integer shows number of months the customer has been with the service. 
Monthly charges is a float showing the charges in decimal values. 
Total charges is an object datatype, however it is most probably a float as the values are floats. We will have to check if all the values are float or some of the values are in text as well. 


Next, we need to explore these features to understand how each of them affects whether customers stop using the service or not, and which feature influence this decision the most. The goal is to find out why customers decide to leave the service. 

In [6]:
with pd.option_context('display.max_rows', 5, 'display.max_columns', None):
    display(df)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
...          ...     ...            ...     ...        ...     ...   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  \
0              No  No phone service             DSL             No   
1             Yes                No             DSL            Yes   
...           ...               ...             ...            ...   
7041          Yes               Yes     Fiber optic             No   
7042          Yes                No     Fiber optic            Yes   

     OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  \
0             Yes               No          No          No              No   
1              No              Yes          No          No              No   
...           ...              ...         ...         ...             ...   
7041           No               No          No          No              No   
7042           No              Yes         Yes         Yes             Yes   

            Contract PaperlessBilling              PaymentMethod  \
0     Month-to-month              Yes           Electronic check   
1           One year               No               Mailed check   
...              ...              ...                        ...   
7041  Month-to-month              Yes               Mailed check   
7042        Two year              Yes  Bank transfer (automatic)   

      MonthlyCharges TotalCharges Churn  
0              29.85        29.85    No  
1              56.95       1889.5    No  
...              ...          ...   ...  
7041           74.40        306.6   Yes  
7042          105.65       6844.5    No  

[7043 rows x 21 columns]

In [27]:
df['Churn'].value_counts(normalize=True).mul(100).round(2)


Churn
No     73.42
Yes    26.58
Name: proportion, dtype: float64

I tried to check why the 'TotalCharges' column was identified as an object instead of a float, which was rather expected. I searched for any non-numerical values or null entries, but didn't find any. Despite this the conversion wasn't working. I then thought there might be spaces before or after the integers, so i used the strip method to remove them. I also implemented a further cleaning step to eliminate any empty subsets that might have remained after the stripping process. After that, the conversion to float was finally successful. 

In [8]:
df['TotalCharges'] = df['TotalCharges'].str.strip()
df = df[df['TotalCharges'] != '']
df['TotalCharges'] = df['TotalCharges'].astype(float)
df['TotalCharges'].value_counts()

TotalCharges
20.20      11
19.75       9
20.05       8
19.90       8
19.65       8
           ..
6849.40     1
692.35      1
130.15      1
3211.90     1
6844.50     1
Name: count, Length: 6530, dtype: int64

In [9]:
df['TotalCharges'].dtype

dtype('float64')

*Gender and Seniorcitizen:*

In [10]:
df.groupby(['gender', 'SeniorCitizen']).agg({
    'Churn': lambda x: (x == 'Yes').sum() / len(x) * 100,
    'MonthlyCharges': 'mean', 
    'TotalCharges' : 'mean'
})

Churn  MonthlyCharges  TotalCharges
gender SeniorCitizen                                         
Female 0              23.979417       62.119811   2174.707256
       1              42.253521       81.109771   2839.934067
Male   0              23.327731       61.656101   2187.343126
       1              41.114983       78.544425   2781.304355

The dataset includes balanced representation of genders. Half of the customers are females and the other half are males.  

Senior citizens (Male and Female) have higher churn proportion than non-senior citizens. Hence, senior citizens are more likely to churn.
The pattern is the same for females and males, same proportions of churning. 

*Gender and Partner*

In [11]:
df.groupby(['gender', 'Partner']).agg({
    'Churn': lambda x: (x == 'Yes').sum() / len(x) * 100,
    'MonthlyCharges': 'mean',
    'TotalCharges': 'mean'
})

Churn  MonthlyCharges  TotalCharges
gender Partner                                         
Female No       34.444444       63.874111   1627.546056
       Yes      18.954248       66.652525   2984.415508
Male   No       31.538880       60.079500   1543.277923
       Yes      20.467836       69.020585   3079.372164

The dataframe above reveals that customers with partners are less likely to churn, where the percentage of customer who churned is much less for those who have partners (Roughly 18-20%), however for those who don't have a partner it increments to 30-35%. 
Also, male customers with partner have the highest monthly and total charges.  

*Dependents*

In [12]:
df.groupby(['Dependents']).agg({
    'Churn' : lambda x: (x == 'Yes').sum()/len(x)*100, 
    'MonthlyCharges' : 'mean', 
    'TotalCharges' : 'mean'
})

Churn  MonthlyCharges  TotalCharges
Dependents                                         
No          31.279140       67.002808   2187.709254
Yes         15.531205       59.617032   2507.955669

30% of the customer who don't have dependents churned, however only 15% of those with dependents did. Thus, Customers with dependents are less likely to churn compared to those without dependents. 

*Tenure*

As the tenure have various values, it may be best to create bins for tenure, in order to get useful insights from it.

In [13]:
df['tenure'].max()

72

In [28]:
df['TenureRange'] = pd.cut(df['tenure'], 
                           bins=[0, 12, 24, 36, 48, 60, 72], 
                           labels=['0-12', '13-24', '25-36', '37-48', '49-60', '61-72'])
#Now we can group by the ranges we used
df.groupby('TenureRange').agg({
    'Churn': lambda x: (x == 'Yes').sum() / len(x) * 100,
    'MonthlyCharges': 'mean',
    'TotalCharges': 'mean'
})

Churn  MonthlyCharges  TotalCharges
TenureRange                                         
0-12         47.678161       56.172023    276.621563
13-24        28.710938       61.357275   1126.257520
25-36        21.634615       65.575481   1990.199279
37-48        19.028871       66.318241   2827.473163
49-60        14.423077       70.550781   3848.132572
61-72         6.609808       75.952701   5180.669829

There is a decreasing trend of churn proportions among the different tenure range. For 0-12 months range, the churn proportion is the highest (~ 50% ), suggesting that new customers are more likely to churn, however churn proportions decrease as tenure increase, indicating that customers who have been with the service longer are less likely yo churn.

Tenure does have a remarkable influence on the churn

*Phone Service, Multiple Lines*

In [15]:
df.groupby(['PhoneService', 'MultipleLines'])['Churn'].value_counts(normalize=True).mul(100).round(2).unstack()

Churn                             No    Yes
PhoneService MultipleLines                 
No           No phone service  75.00  25.00
Yes          No                74.92  25.08
             Yes               71.35  28.65

This analysis shows that whether the customer has a phone service and whether he has multiple lines doesn't have a siginificant effect on the churn rate, as we can see it is approximately same for all cases

*Internet Services*

In [16]:
df.groupby([ 'InternetService']).agg({
    'Churn' : lambda x: (x == 'Yes').sum()/len(x)*100, 
    'MonthlyCharges' : 'mean', 
    'TotalCharges' : 'mean'
})

Churn  MonthlyCharges  TotalCharges
InternetService                                         
DSL              18.998344       58.088017   2119.789259
Fiber optic      41.892765       91.500129   3205.304570
No                7.434211       21.076283    665.220329

The internet service has a huge impact on the churn rate among customers. For those with DSL service the churn rate is much lower (below 20%) than those with Fiber optic (40%) where it is doubled. This is most probably due to the fact that the fiber optics has a much higher monthly/total charge

*Online Security, Online Backup, Device Protection, Tech Support*

In [17]:
df.groupby(['OnlineSecurity', 'OnlineBackup'])['Churn'].value_counts(normalize=True).mul(100).round(2).unstack()

Churn                                       No    Yes
OnlineSecurity      OnlineBackup                     
No                  No                   51.85  48.15
                    Yes                  68.97  31.03
No internet service No internet service  92.57   7.43
Yes                 No                   80.27  19.73
                    Yes                  89.40  10.60

In [18]:
df.groupby(['OnlineSecurity','DeviceProtection'])['Churn'].value_counts(normalize=True).mul(100).round(2).unstack()

Churn                                       No    Yes
OnlineSecurity      DeviceProtection                 
No                  No                   52.95  47.05
                    Yes                  67.02  32.98
No internet service No internet service  92.57   7.43
Yes                 No                   79.93  20.07
                    Yes                  89.80  10.20

In [19]:
df.groupby(['OnlineSecurity', 'TechSupport'])['Churn'].value_counts(normalize=True).mul(100).round(2).unstack()

Churn                                       No    Yes
OnlineSecurity      TechSupport                      
No                  No                   51.04  48.96
                    Yes                  77.65  22.35
No internet service No internet service  92.57   7.43
Yes                 No                   78.67  21.33
                    Yes                  90.97   9.03

In [20]:
df.groupby(['OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport'])['Churn'].value_counts(normalize=True).mul(100).round(2).unstack()

Churn                                                                               No  \
OnlineSecurity      OnlineBackup        DeviceProtection    TechSupport                  
No                  No                  No                  No                   43.33   
                                                            Yes                  69.01   
                                        Yes                 No                   53.59   
                                                            Yes                  76.25   
                    Yes                 No                  No                   58.39   
                                                            Yes                  80.68   
                                        Yes                 No                   65.04   
                                                            Yes                  84.71   
No internet service No internet service No internet service No internet service  92.57   
Yes                 No                  No                  No                   70.13   
                                                            Yes                  84.43   
                                        Yes                 No                   81.71   
                                                            Yes                  89.26   
                    Yes                 No                  No                   80.37   
                                                            Yes                  89.91   
                                        Yes                 No                   86.49   
                                                            Yes                  94.67   

Churn                                                                              Yes  
OnlineSecurity      OnlineBackup        DeviceProtection    TechSupport                 
No                  No                  No                  No                   56.67  
                                                            Yes                  30.99  
                                        Yes                 No                   46.41  
                                                            Yes                  23.75  
                    Yes                 No                  No                   41.61  
                                                            Yes                  19.32  
                                        Yes                 No                   34.96  
                                                            Yes                  15.29  
No internet service No internet service No internet service No internet service   7.43  
Yes                 No                  No                  No                   29.87  
                                                            Yes                  15.57  
                                        Yes                 No                   18.29  
                                                            Yes                  10.74  
                    Yes                 No                  No                   19.63  
                                                            Yes                  10.09  
                                        Yes                 No                   13.51  
                                                            Yes                   5.33

After checking for relationships between all of these 4 attributes with the churn, i got several insights:
1. Customers with no internet services, have lower churn rates.
2. Customers with the combination of all of these 4 services has the lowest churn rates, it decreases from 56.6% (For customers with internet service, but none of these services) to 5.33%(for customers with internet service and all of the 4 services) which is a huge decrease.
3. For customers with combinations of at least 2 of these services, the churn rate drops by half, however customers with individual services have higher rate.


*StreamingTv and StreamingMovies*

In [21]:
df.groupby(['StreamingTV', 'StreamingMovies'])['Churn'].value_counts(normalize=True).mul(100).round(2).unstack()

Churn                                       No    Yes
StreamingTV         StreamingMovies                  
No                  No                   65.54  34.46
                    Yes                  68.81  31.19
No internet service No internet service  92.57   7.43
Yes                 No                   68.19  31.81
                    Yes                  70.55  29.45

The presence or absence of Streaming TV and Streaming Movies seems to influence the churn rate to a limited extent. There isn't much of a different between their presence and absence. 

*Contract*

In [22]:
#Churn percentage by contract:
df.groupby('Contract')['Churn'].value_counts(normalize=True).mul(100).unstack()

Churn                  No        Yes
Contract                            
Month-to-month  57.290323  42.709677
One year        88.722826  11.277174
Two year        97.151335   2.848665

In [23]:
df.groupby('Contract').agg({
    'Churn': lambda x: (x == 'Yes').sum() / len(x) * 100,
    'MonthlyCharges': 'mean', 
    'TotalCharges': 'mean'
})

Churn  MonthlyCharges  TotalCharges
Contract                                               
Month-to-month  42.709677       66.398490   1369.254581
One year        11.277174       65.079416   3034.683084
Two year         2.848665       60.872374   3728.933947

The contract variable has a very huge influence, customers with month-to-month contracts are more likely to churn(42%). The percentage is way less for those with one-year contract (11%) and two year (3% only)

*PaperlessBilling, PaymentMethod*

In [24]:
df.groupby(['PaperlessBilling', 'PaymentMethod'])['Churn'].value_counts(normalize=True).mul(100).unstack()

Churn                                              No        Yes
PaperlessBilling PaymentMethod                                  
No               Bank transfer (automatic)  88.973966  11.026034
                 Credit card (automatic)    89.984351  10.015649
                 Electronic check           67.255217  32.744783
                 Mailed check               86.406744  13.593256
Yes              Bank transfer (automatic)  79.077615  20.922385
                 Credit card (automatic)    80.952381  19.047619
                 Electronic check           50.229621  49.770379
                 Mailed check               72.671756  27.328244

- Customers with Paperless billing have higher percentages of churn than those without. (The percentage is doubled)
- Payment methods: Customers who use electronic check are more prone to churn, followed by mailed check. Customers with Automatic payment methods, such as bank transfer or credit card, have the least percentage.
- Hence payment methods such as Electronic check or mailed check might be influencing the customer's decision to leave the service, however the automatic methods might have a better contribution to the customer's retention to the service.

*Monthly Charges and Total Charges*

In [25]:
df.groupby('Churn').agg({
    'MonthlyCharges': ['mean', 'std'],
    'TotalCharges': ['mean', 'std']
})

MonthlyCharges            TotalCharges             
                mean        std         mean          std
Churn                                                    
No         61.307408  31.094557  2555.344141  2329.456984
Yes        74.441332  24.666053  1531.796094  1890.822994

Customers who churned tend to have higher monthly charges but lower total charges. Customers with higher monthly charges might be more sensitive to prices changes, which might influence their decision to churn. Lower total charges for churned customers might suggest that they're leaving before reaching higher total charge amounts. 

**Conclusion**

Customer churn is impacted by several factors, including contract length, payment methods, internet service quality, tenure date, monthly and total charges, and other variables. By encouraging longer contract commitments, enhancing payment methods(relying more on automatic payment methods), implementing competitive monthly pricing, businesses can effectively reduce churn rates and attract customers to retain their services